libyul/backends/wasm/EVMToEWasmTranslator.cpp

```diff
@@ -314,20 +696,68 @@ function shr(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
 	}
 }
 function sar(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	let lz := i64.and(i64.shl(y1, 63), 1)
+	if i64.eqz(i64.and(lz, 0xffffffffffffffff)) {
+		z1, z2, z3, z4 := shr(x1, x2, x3, x4, y1, y2, y3, y4)
+	}
+	if i64.and(lz, 0xffffffffffffffff) {
+		if i64.ge_u(x4, 256) {
+			z1 := 0xffffffffffffffff
+			z2 := 0xffffffffffffffff
+			z3 := 0xffffffffffffffff
+			z4 := 0xffffffffffffffff
+		}
+		if i64.lt_u(x4, 256) {
+			let sr1, sr2, sr3, sr4 := shr(x1, x2, x3, x4, y1, y2, y3, y4)
+			let mo1, mo2, mo3, mo4 := sub(0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, x1, x2, x3, x4)
+			let sl1, sl2, sl3, sl4 := shl(mo1, mo2, mo3, mo4, y1, y2, y3, y4)
+			z1, z2, z3, z4 := or(sr1, sr2, sr3, sr4,sl1, sl2, sl3, sl4)
+		}
+	}
 }
 function addmod(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	let carry
+	let r1
+	let r2
+	let r3
+	let r4
+	r4, carry := add_carry(x4, y4, 0)
+	r3, carry := add_carry(x3, y3, carry)
+	r2, carry := add_carry(x2, y2, carry)
+	r1, carry := add_carry(x1, y1, carry)
+
+	let z5
+	z1, z2, z3, z4, z5 := mod320(carry, r1, r2, r3, r4, 0, y1, y2, y3, y4)
 }
 function mulmod(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	let r1, r2, r3, r4, r5, r6, r7, r8 := mul_256x256_512(x1, x2, x3, x4, y1, y2, y3, y4)
+	let t1
+	let t2
+	let t3
+	let t4
+	t1, t2, t3, t4, z1, z2, z3, z4 := mod512(r1, r2, r3, r4, r5, r6, r7, r8, 0, 0, 0, 0, y1, y2, y3, y4)
 }
 function signextend(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	if i64.lt_u(y4, 31) {
+		let signBit := i64.add(i64.mul(i64.shr_u(i64.shl(y4, 32), 32), 8), 7)
+		let sm1, sm2, sm3, sm4 := shl(0, 0, 0, 1, 0, 0, 0, signBit)
+		let vm1, vm2, vm3, vm4 := sub(sm1, sm2, sm3, sm4, 0, 0, 0, 1)
+		let ys1, ys2, ys3, ys4 := and(y1, y2, y3, y4, sm1, sm2, sm3, sm4)
+		let isZero := iszero(ys1, ys2, ys3, ys4)
+
+		if isZero {
+			z1, z2, z3, z4 := and(y1, y2, y3, y4, vm1, vm2, vm3, vm4)
+		}
+		if i64.eqz(isZero) {
+			let nvm1, nvm2, nvm3, nvm4 := not(vm1, vm2, vm3, vm4)
+			z1, z2, z3, z4 := and(y1, y2, y3, y4, nvm1, nvm2, nvm3, nvm4)
+		}
+	}
+}
+function u256_to_i128(x1, x2, x3, x4) -> v1, v2 {
+	if i64.ne(0, i64.or(x1, x2)) { invalid() }
```

Comment:
What about the highest bit of x3?

Reply:
I think the name of this function is wrong. It should be u256_to_u128. Its result is always unsigned int 256 bits. OK?